### ISSUE TITLE
Feature/Help request: conditional line-in to pipe

### ISSUE THREAD

**Issue by cdlenfert** [ejurgensen/forked-daapd#632]
I'm attempting to do something similar to [1liminal1 in this closed issue](https://github.com/ejurgensen/forked-daapd/issues/371). I have a cheap bluetooth receiver with auxiliary output (most likely analog). It would be cool to add bluetooth connectivity as an option to my forked-daapd server, and right now I can stream to the BT receiver, which is connected to my USB DAC's mic input, capture that audio and pipe it into my forked-daapd library. FD picks up the pipe, but only once I manually enter a command to capture the audio and send it to the pipe.
`arecord --device=plughw:CARD=Device,DEV=0 -f cd -t wav > /home/share/blueplay`

My goal, since this is installed in our office, is to make it friendlier for non-iPhone users, or in times when our network gets jacked up, let bluetooth be a fallback option. Currently I'm using Shairport-sync as the airplay receiver that pipes to forked-daapd.

Here's what I'm wondering:
- could there be a way to not autostart 'bluepipe', but still be able to manually start it via the remote app, or web interface?
- (or better yet) is it possible to only record data when the bluetooth device is sending audio to the mic?
- and if either of these solutions is viable, how might I go about keeping the arecord process running full time through reboots?

Some thoughts on the conf file:
`	# Watch named pipes in the library for data and autostart playback when
	# there is data to be read. To exclude specific pipes from watching,
	# consider using the above _ignore options.
#	pipe_autostart = true
}`
this points me to the _ignore options, but the ignore options look like they'll exclude the pipe from showing in my library
`	# File types the scanner should ignore
	# Non-audio files will never be added to the database, but here you
	# can prevent the scanner from even probing them. This might improve
	# scan time. By default .db, .ini, .db-journal, .pdf and .metadata are
	# ignored.
#	filetypes_ignore = { ".db", ".ini", ".db-journal", ".pdf", ".metadata" }`

**Comment by ejurgensen** [comment]
I think shairport-sync has the option "run_this_before_play_begins" which could perhaps serve you in this matter. You should then set pipe_autostart to false, and then instead achieve autostart by having shairport-sync start forked-daapd playback via the option. Something like this:

`run_this_before_play_begins="mpc add file:/home/share/shairportpipe && mpc play"`

I'm not that strong with mpc commands, so the above might not be spot on. I expect you will need to experiment a bit. Also haven't ever actually tried this shairport-sync option, I don't know if compounding the commands with && will work.

This solution would require manual start of the Bluetooth pipe, and a better solution would be to have that autostart as well. I imagine that would require a service that listens on the aux input all the time, and when the volume reaches a certain level it kicks of "arecord > pipe". Don't really know how to do that exactly, maybe google does :-)

If you find a solution please post it here, might help other people.

**Comment by xiubert** [comment]
This may also be helpful:

https://github.com/b-fitzpatrick/cpiped

I used to use it w/ snapcast before switching to forked-daapd + shairport-sync

Considering using it again for mic and bluetooth input

**Comment by cdlenfert** [comment]
@xiubert Awesome! I did the same and it's working great. Pics up the line-in when audio is playing and feeds the pipe, and when audio stops clears it so other (airplay) pipes can be played. Very very cool. I'd love to create a systemd "unit" for cpiped. That's also a bit beyond my skill set :)

**Comment by xiubert** [comment]
I use something like below on Arch.  I assume if your distro uses systemd it should work for you. Edit accordingly. 
systemctl start cpiped.service, systemctl status cpiped.service, if all good: systemctl enable cpiped.service

cat /etc/systemd/system/cpiped.service:
[Unit]
Description=cpiped
After=sound.target

[Service]
Type=forking
KillMode=none
User=forkedDAAPDusr
ExecStart=/usr/local/bin/cpiped -d "hw:0,0" -s /home/forkedDAAPDusr/soundDetect.sh -e /home/forkedDAAPDusr/soundAbsence.sh -D /home/forkedDAAPDusr/detectionPipe
ExecStop=/usr/bin/killall -9 cpiped
WorkingDirectory=/home/forkedDAAPDusr

[Install]
WantedBy=multi-user.target

**Comment by cdlenfert** [comment]
@xiubert Worked great and came back up on its own upon reboot. Thank you!

**Comment by ejurgensen** [comment]
It would be great if you would add a wiki page on this @xiubert!

**Comment by sebkacz** [comment]
Just wanted to add (for anyone else reading this), my Raspberry Pi Zero would work well the first time it detected sound, but would then jitter any subsequent times after.
My setup is slightly different however as I have a record player plugged into the USB input of my Pi.

I'm not sure what was causing the jitter, but I think the second pipe was still receiving data somehow, but way out of my scope of knowledge.
Restarting the Pi would fix it.

My fix was by creating a hard link between the detection pipe and the forked-daapd pipe on sound detection, and deleting that hard link on sound absence.

cpiped.service
`[Unit]
Description=cpiped
After=sound.target

[Service]
Type=forking
KillMode=none
User=pi
ExecStart=/usr/local/bin/cpiped -d "hw:1,0" -b 2 -s /home/pi/soundDetect.sh -e /home/pi/soundAbsence.sh -D /home/pi/detectionPipe
ExecStop=/usr/bin/killall -9 cpiped
WorkingDirectory=/home/pi

[Install]
WantedBy=multi-user.target`

soundDetect.sh
`#!/bin/bash
sudo ln /home/pi/detectionPipe /var/media/recordPlayer`

soundAbsence.sh
`#!/bin/bash
sudo rm -rf /var/media/recordPlayer`

I just need to work out now how to handle the rejection of the AppleTV turning off (it has a little bit of a freakout), and to reconnect on input (currently selects the first AirPlay device being my HomePod)

**Comment by ayman** [comment]
@sebkacz > My fix was by creating a hard link between the detection pipe and the forked-daapd pipe on sound detection, and deleting that hard link on sound absence.

Does it have to be a hard link? Would a symbolic work (and be friendlier in general)?

**Comment by ayman** [comment]
> It would be great if you would add a wiki page on this @xiubert!

Since I did this setup from scratch and took notes, I made the [how to page on the Wiki](https://github.com/ejurgensen/forked-daapd/wiki/Making-an-Analog-In-to-Airplay-RPi-Transmitter).

**Comment by ejurgensen** [comment]
Very good instructions - thanks!

**Comment by xiubert** [comment]
Those are slick!  Thanks for doing that.  I'll likely be referring to them instead of my scattered notes when I get to it again.

**Comment by cdlenfert** [comment]
@ayman nice writeup indeed! This setup has been working well for me. I'm now trying to do something similar on a shairport-sync endpoint device. The idea is a bit different in that I have an Echo Flex (or any analog output device really) that should always be sending Mic input out through the soundcard's output. This way I can ask the Echo questions, playback music by voice, etc, but then when I want to airplay to it, that works too. I've got it working but for some reason my arecord process just stops after a certain amount of time. I'll keep working on it, but I definitely like the idea of my source AND endpoints being able to use AUX input.

**Comment by ayman** [comment]
Thanks everyone. This thread was the only place that really revealed the solution; glad you liked the wiki writeup!

> some reason my arecord process just stops after a certain amount of time

So `arecord -d 0 ...` should just record forever; I actually had it die/timeout when my HiFiBerry DAC+ADC wasn't configured to clock slave mode for whatever reason. Maybe check the `dt-overlay` settings of your ADC device if on a RPi which also needed an `rpi-update` (roadblocked me for a week on that one which is why I dropped it at the bottom of the wiki).

**Comment by GrandMaster0** [comment]
First of all, thanks for the great writeup! I want my record player to play music on my HomePod and this seems to be the perfect solution. I managed to play the example mp3 file over forked-daapd to the HomePod and "arecord" worked as well (tested with music from my iPad over the audio cable to the raspi). Now with cpiped it kind of works but it is very pitched up... Does someone know what might be the problem here?

**Comment by xiubert** [comment]
> First of all, thanks for the great writeup! I want my record player to play music on my HomePod and this seems to be the perfect solution. I managed to play the example mp3 file over forked-daapd to the HomePod and "arecord" worked as well (tested with music from my iPad over the audio cable to the raspi). Now with cpiped it kind of works but it is very pitched up... Does someone know what might be the problem here?

likely a sample rate issue

**Comment by GrandMaster0** [comment]
yes, but how/where could I fix it?

**Comment by Aeroform-se** [comment]
Thank you so much for that awesome writeup @ayman . Got it working perfectly even though it was my first ever rpi/linux project! I don't know if I did something wrong, but I had to create the `/home/pi/.audioDetectPipe` pipe before it would work (in case that step simply has been forgotten in the wiki?).

On top of this I added quite a lot of extra stuff to handle my quirky usb-equipped vinyl player:
- udev rule -> systemd to launch and kill cpiped when turning on/off my record player.
- bash script to handle different errors in starting cpiped for my particular dac/player (some errors such as "clock source 41 is not valid" causing cpiped to fail on first try but working on second/third, seems to be a quite normal issue for certain high-res audio DACs).
- adding a soft volume control to alsamixer as the output from my record player was on the low side.

Please let me know if you want me to contribute to the wiki entry, I have everything noted down already as well as (really crappy I guess) code examples.

**Comment by ayman** [comment]
> Please let me know if you want me to contribute to the wiki entry, I have everything noted down already as well as (really crappy I guess) code examples.

Please contribute! It's a wiki. I can't recall if i made that `/home/pi/.audioDetectPipe` before or not (or if my dir had the right permissions to just let `cpiped` make it already).   I'm about to hook it up to my turntable, so  please add notes and fixes. Thanks @Aeroform-se!

**Comment by tonysprenk** [comment]
Hi all, I got everything to work from the wiki instructions but for some reason im not able to play the aux file in forked-daapd. It worked before configuring cpiped. after configuring cpiped the file is there but when I press play nothing happens even though my turntable is putting out audio. what can I do to troubleshoot?

**Comment by mixographer** [comment]
I have a question about how you get the fifo to go in a reasonable place in the sorting list? My fifo winds up in 'Unknown artist,' Which means every time I want to play the truntable, I need to scroll all the way down, and then tap into Unknown and then find my fifo. Can anyone tell me how I can make it an album or an artisit so it will sort up at the top?

**Comment by mixographer** [comment]
In answer to my last comment, I see it is a matter of adding these config options to the conf file:

CFG_STR("name_unknown_artist", "Unknown artist", CFGF_NONE),
CFG_STR("name_unknown_album", "Unknown album", CFGF_NONE),

like 

name_unknown_artist = "Analog Turntable" 

or something. Still trying to understand how it all works

**Comment by ejurgensen** [comment]
Yes, it isn't really documented, but you should be able to change the name by adding `name_unknown_artist = "Analog Turntable"` to the library section of the config file. After you have done that and restarted the server, you will probably also have to trigger a rescan of the fifo which you can do with the command `touch /path/to/pipe`. If that doesn't work then try a rescan (web interface > Update library > Rescan metadata for unmodified files).

Any particular reason you aren't using autostart of the pipe?

**Comment by cdlenfert** [comment]
@mixographer I'd also highly recommend autostarting the pipe. In the case of using cpiped it's important to stop remove the pipe when it's not receiving audio, but that's all covered in the wiki I belive.

**Comment by noahals99** [comment]
This is great, works wonders! I do have one issue if anyone would possibly know a solution. The audio seems to jump around at times, it will skip ahead a couple second every 35 second or so. Another issue is that the audio seems to be pitched down a bit. Any help would be appreciated, thanks!

**Comment by snacques** [comment]
@noahals99 if you (or anyone else landing here via search) have been unable to resolve these symptoms, I experienced them too and was able to fix them on my instance.

My problem was that my mic dongle (Plugable USB-AUDIO) is fixed at a sampling rate of 48kHz. Whereas Owntone, to the best of my knowledge, is fixed to consume from named pipes at 44.1kHz. So I used sox to downsample.

arecord example:
```
arecord -D hw:CARD=Device,DEV=0 -f dat -d 0 -t raw | sox -t raw -r 48000 -e signed -b 16 -c 2 - -t raw -r 44100 -e signed -b 16 -c 2 - > /home/pi/music/AUX
```


You'll need to replace your own CARD and DEV ids. Also, you'll need to confirm your capture card sampling rate, etc. with `arecord --dump-hw-params -D hw:1,0` substituting your "hw" ids. You need to be explicit in the sox command, but arecord is more forgiving.


If that works, use this in soundDetect.sh for the cpiped solution, with your input card details:
```
#!/bin/bash
sox -t raw -r 48000 -e signed -b 16 -c 2 /home/pi/.audioDetectPipe \
    -t raw -r 44100 -e signed -b 16 -c 2 - > /home/pi/music/AUX &

echo $! > /tmp/catPipe.pid
```

I will update the [wiki instructions](https://github.com/owntone/owntone-server/wiki/Making-an-Analog-In-to-Airplay-RPi-Transmitter) with these details.

**Comment by ale275** [comment]
Hi

don't know if useful for anyone but I tried to put together everything from wiki plus some more and I created this small project [audio2pipe](https://github.com/ale275/audio2pipe) that relies on my forked version of [cpiped](https://github.com/ale275/cpiped)

I made template for systemd service
Automatic output selection
Support for different sample rates, including clan killing

and some other stuff
